src/libraries/Common/src/System/Threading/Tasks/TaskTimeoutExtensions.cs

```diff
@@ -0,0 +1,46 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+namespace System.Threading.Tasks
+{
+    /// <summary>
+    /// Task timeout helper based on https://devblogs.microsoft.com/pfxteam/crafting-a-task-timeoutafter-method/
+    /// </summary>
+    internal static class TaskTimeoutExtensions
+    {
+        public static Task WithCancellation(this Task task, CancellationToken cancellationToken)
+        {
+            if (task is null)
+            {
+                throw new ArgumentNullException(nameof(task));
+            }
+
+            if (task.IsCompleted || !cancellationToken.CanBeCanceled)
+            {
+                return task;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            return WithCancellationCore(task, cancellationToken);
+
+            static async Task WithCancellationCore(Task task, CancellationToken cancellationToken)
+            {
+                var tcs = new TaskCompletionSource();
+                using IDisposable _ = cancellationToken.UnsafeRegister(static s => ((TaskCompletionSource)s!).TrySetResult(), tcs);
```

Comment:
This is going to box the `CancellationTokenRegistration` struct into an `IDisposable`.  It should either be:
```suggestion
                using CancellationTokenRegistration _ = cancellationToken.UnsafeRegister(static s => ((TaskCompletionSource)s!).TrySetResult(), tcs);
```
or
```suggestion
using (cancellationToken.UnsafeRegister(static s => ((TaskCompletionSource)s!).TrySetResult(), tcs))
```
with the appropriate braces for the code below.  Unfortunately the new using statement doesn't yet support discards.